cpp/Platform.Delegates/Platform.Delegates.Delegate.h

```diff
@@ -16,54 +14,134 @@ namespace Platform::Delegates
+            }
+            return *this;
+        }
+
+        Delegate &operator=(Delegate&& other) noexcept {
+            if (this != &other)
+            {
+                this->simpleFunction = std::move(other.simpleFunction);
+                this->memberMethod = std::move(other.memberMethod);
+                this->complexFunction = std::move(other.complexFunction);
+            }
+            return *this;
+        }
+
+        virtual ReturnType operator()(Args... args)
+        {
+            if (simpleFunction)
+            {
+                return simpleFunction(args...);
+            }
+            if (memberMethod)
+            {
+                return (*memberMethod)(args...);
+            }
+            if (complexFunction)
+            {
+                return (*complexFunction)(args...);
+            }
+
+            throw std::bad_function_call{};
+        }
+
+        virtual bool operator==(const Delegate &other) const
+        {
+            if (simpleFunction && other.simpleFunction)
+            {
+                return simpleFunction == other.simpleFunction;
+            }
+            if (memberMethod && other.memberMethod)
+            {
+                return *memberMethod == *other.memberMethod;
+            }
+            if (complexFunction && other.complexFunction)
+            {
+                return AreFunctionsEqual(*complexFunction, *other.complexFunction);
+            }
+            return false;
+        }
+
+        virtual bool operator!=(const Delegate &other) const
+        {
+            return !(*this == other);
+        }
+    private:
         class MemberMethodBase
         {
         public:
             virtual ReturnType operator()(Args... args) = 0;
             virtual bool operator== (const MemberMethodBase &other) const = 0;
             virtual bool operator!= (const MemberMethodBase &other) const = 0;
+            virtual ~MemberMethodBase() = default;
         };
 
         template <typename Class>
         class MemberMethod : public MemberMethodBase
         {
-            std::shared_ptr<Class> object;
-            ReturnType(Class:: *member)(Args...);
-
         public:
+            MemberMethod(std::shared_ptr<Class> object, ReturnType(Class:: *method)(Args...)) noexcept
+                    : object(std::move(object)), method(method) { }
 
-            MemberMethod(std::shared_ptr<Class> object, ReturnType(Class:: *member)(Args...)) : object(object), member(member) { }
-
-            virtual ReturnType operator()(Args... args) override
+            ReturnType operator()(Args... args) override
             {
-                return (this->object.get()->*this->member)(std::forward<Args>(args)...);
+                return (*object.*method)(args...);
             }
 
-            virtual bool operator== (const MemberMethodBase &other) const override
+            bool operator== (const MemberMethodBase &other) const override
```

Comment:
Is it more efficient if the `virtual` keyword is removed?

Reply:
Yes, for example in this case:
```c++
void foo() {
    MemberMethod m1, m2;
    /* some code */
    m1 == m2; 
}
```
Static function call, instead of pulling the function out of the vtable. This is faster, because unnecessary indirection disappears, and optimizations can be made, such as function inlining

Reply:
Super.